collector/receiver/otelarrowreceiver/internal/arrow/arrow.go

```diff
@@ -350,13 +364,52 @@ func (r *Receiver) anyStream(serverStream anyStreamServer, method string) (retEr
 		}
 	}()
 
+	doneCtx, doneCancel := context.WithCancel(streamCtx)
+	streamErrCh := make(chan error, 2)
+	pendingCh := make(chan batchResp, runtime.NumCPU())
+
+	go func() {
+		err := r.srvReceiveLoop(doneCtx, serverStream, streamErrCh, pendingCh, method, ac)
+		streamErrCh <- err
+	}()
+
+	// WG is used to ensure main thread only returns once sender is finished sending responses for all requests.
+	var senderWG sync.WaitGroup
+	senderWG.Add(1)
+	go func() {
+		err := r.srvSendLoop(doneCtx, serverStream, pendingCh)
+		streamErrCh <- err
+		senderWG.Done()
+	}()
+
+
+	select {
+	case <-doneCtx.Done(): 
+		senderWG.Wait()
+		return status.Error(codes.Canceled, "server stream shutdown")
+	case retErr = <-streamErrCh:
+		doneCancel()
+		senderWG.Wait()
+		return
+	}
+}
+
+func (r *Receiver) srvReceiveLoop(ctx context.Context, serverStream anyStreamServer, streamErrCh chan<- error, pendingCh chan<- batchResp, method string, ac arrowRecord.ConsumerAPI) (retErr error) {
+	hrcv := newHeaderReceiver(ctx, r.authServer, r.gsettings.IncludeMetadata)
 	for {
+		select {
+		case <-ctx.Done():
+			return status.Error(codes.Canceled, "server stream shutdown")
+		default:
+			r.inFlightWG.Add(1)
```

Comment:
Thanks for the review! Rewrote using some of your suggestions but kept the entire for loop logic in the same `recvOne` function to keep return value simple. I added a defer at the top

```
	defer func() {
		if retErr != nil {
			r.inFlightWG.Done() // not processed
			r.logStreamError(retErr)
		}
	}()
```

I have 1 call to `r.inFlightWg.Add(1)` at the top and  2 calls to `r.inFlightWG.Done` in different conditions, otherwise there would be a race. Let me know if this can still be improved.